extension/memorylimiterextension/memorylimiter.go

```diff
@@ -0,0 +1,38 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package memorylimiterextension // import "go.opentelemetry.io/collector/extension/memorylimiterextension"
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/internal/memorylimiter"
+)
+
+// newMemoryLimiter returns a new memorylimiter extension.
+func newMemoryLimiter(cfg *Config, logger *zap.Logger) (*memoryLimiterExtension, error) {
+	ml, err := memorylimiter.NewMemoryLimiter(cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &memoryLimiterExtension{memLimiter: ml}, nil
+}
+
+func (ml *memoryLimiterExtension) Start(_ context.Context, host component.Host) error {
+	ml.memLimiter.SetBallastSizeIfExtensionExists(host)
+	ml.memLimiter.StartMonitoring()
+	return nil
+}
+
+func (ml *memoryLimiterExtension) Shutdown(_ context.Context) error {
+	return ml.memLimiter.StopMonitoring()
+}
+
+// MustRefuse returns if the caller should deny because memory has reached it's configured limits
+func (ml *memoryLimiterExtension) MustRefuse() bool {
+	return ml.memLimiter.MustRefuse()
+}
```

Comment:
Please also implement https://github.com/open-telemetry/opentelemetry-collector/blob/44fbb84a0fee035891de960b0ad82160e264267a/extension/extension.go#L22 to make sure that the ballast extension is started before the memoey_limiter. Otherwise, we won't get the right ballast value

Reply:
impled

Reply:
I've just realized that this interface actually requires the extension it depends on to be provided. Otherwise, it fails to start. The `memory_ballast` extension is optional and even discouraged. So, we cannot use this approach. I submitted https://github.com/open-telemetry/opentelemetry-collector/pull/9218 to make sure that the ballast is always calculated before the `memory_extension` is started. So we can remove this one. Sorry for the back-and-forths.